blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1066 @@
+        }
+    }
+
+    /// Return Vector of signed messages given a block header for self
+    pub async fn messages_for_blocks(
+        &mut self,
+        blks: &[BlockHeader],
+    ) -> Result<Vec<SignedMessage>, Error> {
+        let mut msg_vec: Vec<SignedMessage> = Vec::new();
+
+        for block in blks {
+            let api = self.api.read().await;
+            let (umsg, mut smsgs) = api.messages_for_block(&block)?;
+            drop(api);
+
+            msg_vec.append(smsgs.as_mut());
+            for msg in umsg {
+                let smsg = recover_sig(self.bls_sig_cache.clone(), msg).await?;
+                msg_vec.push(smsg)
+            }
+        }
+        Ok(msg_vec)
+    }
+
+    /// Return gas price estimate this has been translated from lotus, a more smart implementation will
+    /// most likely need to be implemented
+    pub fn estimate_gas_price(&self, nblocksincl: u64) -> Result<BigInt, Error> {
+        // TODO possibly come up with a smarter way to estimate the gas price
+        let min_gas_price = 0;
+        match nblocksincl {
+            0 => Ok(BigInt::from(min_gas_price + 2)),
+            1 => Ok(BigInt::from(min_gas_price + 1)),
+            _ => Ok(BigInt::from(min_gas_price)),
+        }
+    }
+
+    /// Load local messages into pending. As of  right now messages are not deleted from self's
+    /// local_message field, possibly implement this in the future?
+    pub async fn load_local(&mut self) -> Result<(), Error> {
+        for (key, value) in self.local_msgs.clone() {
+            self.add(&value).await.unwrap_or_else(|err| {
+                if err == Error::SequenceTooLow {
+                    warn!("error adding message: {:?}", err);
+                    self.local_msgs.remove(&key);
+                }
+            });
+        }
+        Ok(())
+    }
+}
+
+/// Remove a message from pending given the from address and sequence
+pub async fn remove(
+    from: &Address,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    sequence: u64,
+) -> Result<(), Error> {
+    let mut pending = pending.write().await;
+
+    let mset = pending
+        .get_mut(from)
+        .ok_or_else(|| Error::InvalidFromAddr)?;
+    mset.msgs.remove(&sequence);
+
+    if mset.msgs.is_empty() {
+        pending.remove(from);
+    } else {
+        let mut max_sequence: u64 = 0;
+        for sequence_temp in mset.msgs.keys().cloned() {
+            if max_sequence < sequence_temp {
+                max_sequence = sequence_temp;
+            }
+        }
+        if max_sequence < sequence {
+            max_sequence = sequence;
+        }
+        mset.next_sequence = max_sequence + 1;
+    }
+    Ok(())
+}
+
+/// Attempt to get a signed message that corresponds to an unsigned message in bls_sig_cache
+async fn recover_sig(
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    msg: UnsignedMessage,
+) -> Result<SignedMessage, Error> {
+    let mut bls_sig_cache = bls_sig_cache.write().await;
```

Comment:
you don't need write access to do this? can also change the bls_sig_cache type to just be a ref to the lru cache because this function does not need to be aware of the lock

Reply:
write was required to call the get method. with read you can't get a mut RwLock, but that doesn't really matter if I switch to a ref to the lru cache